nats.go

```diff
@@ -3089,6 +3194,98 @@ func (nc *Conn) IsConnected() bool {
 	return nc.isConnected()
 }
 
+// drainConnection will run in a separate Go routine and will
+// flush all publishes and drain all active subscriptions.
+func (nc *Conn) drainConnection() {
+	// Snapshot subs list.
+	nc.mu.Lock()
+	subs := make([]*Subscription, 0, len(nc.subs))
+	for _, s := range nc.subs {
+		subs = append(subs, s)
+	}
+	errCB := nc.Opts.AsyncErrorCB
+	drainWait := nc.Opts.DrainTimeout
+	nc.mu.Unlock()
+
+	// for pushing errors with context.
+	pushErr := func(err error) {
+		nc.mu.Lock()
+		if errCB != nil {
+			nc.ach.push(func() { errCB(nc, nil, err) })
+		}
+		nc.mu.Unlock()
+	}
+
+	// Do subs first
+	for _, s := range subs {
+		err := s.Drain()
+		// We will notify about these but continue.
+		if err != nil {
+			pushErr(err)
+		}
+	}
+
+	// Wait for the subscriptions to drop to zero.
+	timeout := time.Now().Add(drainWait)
+	for time.Now().Before(timeout) {
+		if nc.NumSubscriptions() == 0 {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	// Check if we timed out.
+	if nc.NumSubscriptions() != 0 {
+		pushErr(ErrDrainTimeout)
+	}
+
+	// Flip State
+	nc.mu.Lock()
+	nc.status = DRAINING_PUBS
+	nc.mu.Unlock()
+
+	// Do publish drain via Flush() call.
+	err := nc.Flush()
+	if err != nil {
+		pushErr(err)
+		nc.Close()
+		return
+	}
+
+	// Move to closed state.
+	nc.Close()
+}
+
+// Drain will put a connection into a drain state. Publishes after this call
```

Comment:
Should update comment here. Publishes calls do not really fail after this call, but some time after that ;-)
Also, I think it says it (with reference to "Upon completion ..."), but maybe we should make it clear that this call does not block waiting for drain to complete.

Reply:
fixed